compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt

```diff
@@ -43,46 +43,70 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
 
         val calls = callToLabels.map { it.call }
 
-        // Checks if all conditions are CALL EQEQ(tmp_variable, some_constant)
-        if (!areConstComparisons(calls))
+        // Checks if all conditions are CALL EQEQ(_, _)
+        if (!areComparisons(calls))
             return null
 
-        // Subject should be the same for all conditions. Let's pick the first.
-        val subject = callToLabels[0].call.getValueArgument(0)!! as IrGetValue
-
-        // Don't generate repeated cases, which are unreachable but allowed in Kotlin.
-        // Only keep the first encountered case:
-        val cases =
-            callToLabels.map { ValueToLabel((it.call.getValueArgument(1) as IrConst<*>).value, it.label) }.distinctBy { it.value }
-
-        // Remove labels and "then expressions" that are not reachable.
-        val reachableLabels = HashSet(cases.map { it.label })
-        expressionToLabels.removeIf { it.label !in reachableLabels }
-
+        // Determine if all comparisons are between a single temporary variable
+        // and constants of a particular type: CALL EQEQ(tmp, constant)
         return when {
-            areConstIntComparisons(calls) ->
+            // UInts are an exceptional case: the IR contains CALL <unsafe-coerce><UInt, Int> around both arguments
+            // to the comparison. We check that structure in `areConstUIntComparison`.
+            areConstUIntComparisons(calls) -> {
+                val (subject, expressionToLabels, cases) = extractUIntSwitchComponents(callToLabels, expressionToLabels)
+                IntSwitch(
+                    subject,
+                    elseExpression,
+                    expressionToLabels,
+                    cases
+                )
+            }
+            // All remaining switchable types (Int and String) are primitives and can be handled uniformly.
+            !areTmpVariableVSConstantComparisons(calls) ->
```

Comment:
Don't think this is really necessary. It seems at this point you just need to check to see if all RHS are `IrConst`. But to accomplish this, you could just modify `checkTypeSpecifics` similar to what you did for the `UInt` version:

```
val rhs = conditions.map { it.getValueArgument(1) as? IrConst<*> }
if (rhs.any { it == null || !irConstPredicate(it) })
```

Reply:
You are right! I rolled it into the previous check, however.